### PR TITLE
add support for hosted cluster image content source policy. This allows mirroring of release images to other registries besides quay.io

### PR DESCRIPTION
**Pull request by relyt0925**

### CODE REVIEW

**Comment by enxebre** [comment]
/lgtm
/approve
/hold
@ironcladlou do we want to put this into third party instead? Please feel free to hold cancel otherwise.

**Comment by csrwng** [comment]
Given that ImageContentSource policy is not only something that affects machine ignition but also needs to be synced into the guest cluster (so ImageStreams can use it), I think it should be part of the HostedCluster API. Just like it's part of an InstallConfig (https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L117)

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 1]
```diff
@@ -1920,6 +1931,54 @@ func (r *HostedControlPlaneReconciler) reconcileOIDCRouteResources(ctx context.C
 	return err
 }
 
+func (r *HostedControlPlaneReconciler) reconcileImageContentSourcePolicy(ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
+	imageContentSourceCRD := manifests.ImageContentSourcePolicy()
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:1935]: nit: technically not a crd, but a cr... `imageContentSourcePolicy` seems a better name.

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 2]
```diff
@@ -1920,6 +1931,54 @@ func (r *HostedControlPlaneReconciler) reconcileOIDCRouteResources(ctx context.C
...
+	})
+	return err
+}
+
+func reconcileImageContentSourceIgnitionConfig(ignitionConfig *corev1.ConfigMap, serializedImageContentSourcePolicy []byte) error {
+	if ignitionConfig.Data == nil {
+		ignitionConfig.Data = map[string]string{}
+	}
+	ignitionConfig.Data[manifests.IgnitionConfigKey] = string(serializedImageContentSourcePolicy)
+	// TODO: remove when un
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:1974]: unfinished comment

**Review by csrwng — COMMENTED** [review]
@relyt0925 looks good, just some nits

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 2 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: good catch!

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 1 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: changed!

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 2 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: removed!

File: api/v1alpha1/hosted_controlplane.go [hunk 3]
```diff
@@ -81,6 +81,9 @@ type HostedControlPlaneSpec struct {
 	// https://docs.openshift.com/container-platform/4.7/rest_api/config_apis/config-apis-index.html
 	// +kubebuilder:validation:Optional
 	Configuration *ClusterConfiguration `json:"configuration,omitempty"`
+
+	// ImageContentSources lists sources/repositories for the release-image content.
```
Comment (csrwng) [inline api/v1alpha1/hosted_controlplane.go:85]: Add `// +optional` here

Comment (relyt0925) [reply api/v1alpha1/hosted_controlplane.go:85]: added

**Pull request merged**